gui/webdisplay/src/TWebWindowsManager.cxx

```diff
@@ -593,29 +617,58 @@ bool ROOT::Experimental::TWebWindowsManager::Show(ROOT::Experimental::TWebWindow
       is_firefox = prog.Length() > 0;
 
 #ifdef _MSC_VER
-      if (win.IsBatchMode())
+      if (batch_mode)
          // there is a problem when specifying the window size with wmic on windows:
          // It gives: Invalid format. Hint: <paramlist> = <param> [, <paramlist>].
-         exec = gEnv->GetValue("WebGui.FirefoxBatch", "fork: -headless -no-remote $url");
+         exec = gEnv->GetValue("WebGui.FirefoxBatch", "fork: -headless -no-remote $profile $url");
       else
-         exec = gEnv->GetValue("WebGui.FirefoxInteractive", "$prog -width=$width -height=$height $url");
+         exec = gEnv->GetValue("WebGui.FirefoxInteractive", "$prog -width=$width -height=$height $profile $url");
 #else
-      if (win.IsBatchMode())
-         exec = gEnv->GetValue("WebGui.FirefoxBatch", "fork:-headless -no-remote $url");
+      if (batch_mode)
+         exec = gEnv->GetValue("WebGui.FirefoxBatch", "fork:-headless -no-remote $profile $url");
       else
-         exec = gEnv->GetValue("WebGui.FirefoxInteractive", "$prog -width $width -height $height \'$url\' &");
+         exec = gEnv->GetValue("WebGui.FirefoxInteractive", "$prog -width $width -height $height $profile \'$url\' &");
 #endif
+
+      if (is_firefox && (exec.Index("$profile") != kNPOS)) {
+         TString repl;
+
+         const char *ff_profile = gEnv->GetValue("WebGui.FirefoxProfile","");
+         const char *ff_profilepath = gEnv->GetValue("WebGui.FirefoxProfilePath","");
+         Int_t ff_randomprofile = gEnv->GetValue("WebGui.FirefoxRandomProfile", 0);
+         if (ff_profile && *ff_profile) {
+            repl.Form("-P %s", ff_profile);
+         } else if (ff_profilepath && *ff_profilepath) {
+            repl.Form("-profile %s", ff_profilepath);
+         } else if ((ff_randomprofile > 0) || (batch_mode && (ff_randomprofile>=0))) {
+
+            gRandom->SetSeed(0);
+
+            TString ffp, ffd;
+            ffp.Form("root_ff_profile_%d", gRandom->Integer(0x100000));
+            ffd.Form("%s/%s", gSystem->TempDirectory(), ffp.Data());
+
+            repl.Form("-profile %s", ffd.Data());
+            if (!batch_mode) repl.Prepend("-no-remote ");
+
+            gSystem->Exec(Form("%s -no-remote -CreateProfile \"%s %s\"", prog.Data(), ffp.Data(), ffd.Data()));
+
+            rmdir = std::string("$rmdir$") + ffd.Data();
```

Comment:
Please use an RAII object (a struct for which you use the destructor) to make sure that the directory gets removed on all code paths, even if there is an error below. Else we're left with stale temp profile dirs in case e.g. `CreateHttpServer()` fails etc. Once everything is up and running, and you've passed `rmdir` to the window, you can reset the RAII object so this function doesn't delete the directory by the window does.

Reply:
Yes - we should have some handle, which propogated together with the connection. 
But it will be large change and I want to postpone it for next PRs